File: mobile/sharedLogic/src/commonMain/kotlin/org/example/project/AuthClient.kt
package org.example.project

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class UserResponse(
    val id: String,
    val email: String,
    val householdId: String,
)

@Serializable
data class AuthResponse(
    val token: String,
    val user: UserResponse,
)

@Serializable
data class RegisterRequest(
    val email: String,
    val password: String,
    val rememberMe: Boolean = true,
)

@Serializable
data class LoginRequest(
    val email: String,
    val password: String,
    val rememberMe: Boolean = true,
)

@Serializable
data class ErrorMessage(
    val message: String,
)

class AuthException(message: String) : Exception(message)

class AuthClient(
    private val baseUrl: String,
    private val httpClient: HttpClient = createHttpClient(),
    private val tokens: TokenStore = platformTokenStore(),
) {
    fun getStoredToken(): String? = tokens.getToken()

    suspend fun register(
        email: String,
        password: String,
        rememberMe: Boolean = true,
    ): AuthResponse {
        val response =
            httpClient.post("$baseUrl/api/auth/register") {
                contentType(ContentType.Application.Json)
                setBody(RegisterRequest(email, password, rememberMe))
            }
        return storeAuth(response, rememberMe)
    }

    suspend fun login(
        email: String,
        password: String,
        rememberMe: Boolean = true,
    ): AuthResponse {
        val response =
            httpClient.post("$baseUrl/api/auth/login") {
                contentType(ContentType.Application.Json)
                setBody(LoginRequest(email, password, rememberMe))
            }
        return storeAuth(response, rememberMe)
    }

    suspend fun logout() {
        val token = tokens.getToken() ?: return
        val response =
            httpClient.post("$baseUrl/api/auth/logout") {
                header(HttpHeaders.Authorization, "Bearer $token")
            }
        tokens.clearToken()
        if (!response.status.isSuccess() && response.status.value != 401) {
            throw AuthException(readError(response))
        }
    }

    suspend fun me(): UserResponse {
        val token = tokens.getToken() ?: throw AuthException("Not signed in")
        val response =
            httpClient.get("$baseUrl/api/auth/me") {
                header(HttpHeaders.Authorization, "Bearer $token")
            }
        if (!response.status.isSuccess()) {
            if (response.status.value == 401) {
                tokens.clearToken()
            }
            throw AuthException(readError(response))
        }
        return response.body()
    }

    private suspend fun storeAuth(
        response: HttpResponse,
        rememberMe: Boolean,
    ): AuthResponse {
        if (!response.status.isSuccess()) {
            throw AuthException(readError(response))
        }
        val body: AuthResponse = response.body()
        tokens.saveToken(body.token, rememberMe)
        return body
    }

    private val errorJson = Json { ignoreUnknownKeys = true }

    private suspend fun readError(response: HttpResponse): String {
        val text =
            runCatching { response.bodyAsText() }.getOrNull()
                ?: return "Auth failed (${response.status})"
        val parsed =
            runCatching {
                errorJson.decodeFromString(ErrorMessage.serializer(), text)
            }.getOrNull()
        return parsed?.message ?: "Auth failed (${response.status})"
    }

    companion object {
        fun create(
            baseUrl: String = apiBaseUrl(),
            tokens: TokenStore = platformTokenStore(),
        ): AuthClient = AuthClient(baseUrl, createHttpClient(), tokens)
    }
}

File: web/src/api/authClient.ts
import { apiBaseUrl } from "@/config"
import { defaultBrowserTokenStore, type TokenStore } from "@/api/tokenStore"
import type {
  AuthResponse,
  ErrorMessage,
  LoginRequest,
  RegisterRequest,
  UserResponse,
} from "@/api/types"

export class AuthClient {
  private readonly baseUrl: string
  private readonly fetchFn: typeof fetch
  private readonly tokens: TokenStore

  constructor(
    baseUrl: string = apiBaseUrl,
    fetchFn: typeof fetch = globalThis.fetch.bind(globalThis),
    tokens: TokenStore = defaultBrowserTokenStore(),
  ) {
    this.baseUrl = baseUrl
    this.fetchFn = fetchFn
    this.tokens = tokens
  }

  getStoredToken(): string | null {
    return this.tokens.get()
  }

  async register(
    email: string,
    password: string,
    rememberMe: boolean = true,
  ): Promise<AuthResponse> {
    const body: RegisterRequest = { email, password, rememberMe }
    const response = await this.postJson("/api/auth/register", body)
    return this.storeAuth(response, rememberMe)
  }

  async login(
    email: string,
    password: string,
    rememberMe: boolean = true,
  ): Promise<AuthResponse> {
    const body: LoginRequest = { email, password, rememberMe }
    const response = await this.postJson("/api/auth/login", body)
    return this.storeAuth(response, rememberMe)
  }

  async logout(): Promise<void> {
    const token = this.tokens.get()
    if (!token) {
      return
    }
    const response = await this.fetchFn(apiUrl(this.baseUrl, "/api/auth/logout"), {
      method: "POST",
      headers: { Authorization: `Bearer ${token}` },
    })
    this.tokens.clear()
    if (!response.ok && response.status !== 401) {
      throw new Error(await readErrorMessage(response, "Sign out failed"))
    }
  }

  async me(): Promise<UserResponse> {
    const token = this.tokens.get()
    if (!token) {
      throw new Error("Not signed in")
    }
    const response = await this.fetchFn(apiUrl(this.baseUrl, "/api/auth/me"), {
      headers: { Authorization: `Bearer ${token}` },
    })
    if (!response.ok) {
      if (response.status === 401) {
        this.tokens.clear()
      }
      throw new Error(await readErrorMessage(response, "Session check failed"))
    }
    return (await response.json()) as UserResponse
  }

  private async storeAuth(
    response: Response,
    rememberMe: boolean,
  ): Promise<AuthResponse> {
    if (!response.ok) {
      throw new Error(await readErrorMessage(response, "Auth request failed"))
    }
    const body = (await response.json()) as AuthResponse
    if (typeof body.token !== "string" || !body.user?.email) {
      throw new Error("Invalid auth response")
    }
    this.tokens.set(body.token, rememberMe)
    return body
  }

  private postJson(path: string, body: unknown): Promise<Response> {
    return this.fetchFn(apiUrl(this.baseUrl, path), {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    })
  }
}

export function apiUrl(baseUrl: string, path: string): string {
  if (!baseUrl) {
    return path
  }
  return new URL(path, `${baseUrl}/`).toString()
}

async function readErrorMessage(
  response: Response,
  fallback: string,
): Promise<string> {
  try {
    const body = (await response.json()) as ErrorMessage
    if (typeof body.message === "string" && body.message.length > 0) {
      return body.message
    }
  } catch {
    // ignore non-JSON bodies
  }
  return `${fallback} (${response.status})`
}
